KMS encryption context not backwards compatible

In SOPS 2.0 KMS encryption context is stored as a string value in the json output where the old python version used JSON Object to store encryption context.

```sh
❯ jq '.sops.kms | .[].context' encrypted-python
{
  "a": "b",
  "c": "d"
}
❯ jq '.sops.kms | .[].context' encrypted-go
"a:b,c:d"
```

The two outputs are incompatible with each other and reading an old SOPS encrypted file with SOPS 2.0 will cause a stack trace.

```
panic: interface conversion: interface is map[string]interface {}, not string [recovered]
	panic: interface conversion: interface is map[string]interface {}, not string

goroutine 1 [running]:
panic(0x46e220, 0xc420077b40)
	/usr/local/Cellar/go/1.7.4/libexec/src/runtime/panic.go:500 +0x1a1
go.mozilla.org/sops/vendor/gopkg.in/urfave/cli%2ev1.HandleAction.func1(0xc42012f7a8)
	/Users/samihaahtinen/sources/go/src/go.mozilla.org/sops/vendor/gopkg.in/urfave/cli.v1/app.go:473 +0x29e
panic(0x46e220, 0xc420077b40)
	/usr/local/Cellar/go/1.7.4/libexec/src/runtime/panic.go:458 +0x243
go.mozilla.org/sops/json.Store.kmsEntries(0xc420158120, 0x2, 0x2, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x2)
	/Users/samihaahtinen/sources/go/src/go.mozilla.org/sops/json/store.go:316 +0x656
go.mozilla.org/sops/json.Store.UnmarshalMetadata(0xc420121400, 0x740, 0x940, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, ...)
	/Users/samihaahtinen/sources/go/src/go.mozilla.org/sops/json/store.go:279 +0x8f5
go.mozilla.org/sops/json.BinaryStore.UnmarshalMetadata(0xc420121400, 0x740, 0x940, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, ...)
	/Users/samihaahtinen/sources/go/src/go.mozilla.org/sops/json/store.go:61 +0x7c
go.mozilla.org/sops/json.(*BinaryStore).UnmarshalMetadata(0x7397c8, 0xc420121400, 0x740, 0x940, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, ...)
	<autogenerated>:21 +0xb0
main.loadEncryptedFile(0xc4200b6dc0, 0x6fd380, 0x7397c8, 0xc420121400, 0x740, 0x940, 0x0, 0x0, 0x0, 0x0, ...)
	/Users/samihaahtinen/sources/go/src/go.mozilla.org/sops/cmd/sops/main.go:74 +0xa1
main.main.func1(0xc4200b6dc0, 0x0, 0x0)
	/Users/samihaahtinen/sources/go/src/go.mozilla.org/sops/cmd/sops/main.go:227 +0x322
reflect.Value.call(0x44f060, 0x51eaa8, 0x13, 0x4de77e, 0x4, 0xc42012f768, 0x1, 0x1, 0xfe068, 0x4d3800, ...)
	/usr/local/Cellar/go/1.7.4/libexec/src/reflect/value.go:434 +0x5c8
reflect.Value.Call(0x44f060, 0x51eaa8, 0x13, 0xc42012f768, 0x1, 0x1, 0x51e9e8, 0x0, 0x0)
	/usr/local/Cellar/go/1.7.4/libexec/src/reflect/value.go:302 +0xa4
go.mozilla.org/sops/vendor/gopkg.in/urfave/cli%2ev1.HandleAction(0x44f060, 0x51eaa8, 0xc4200b6dc0, 0x0, 0x0)
	/Users/samihaahtinen/sources/go/src/go.mozilla.org/sops/vendor/gopkg.in/urfave/cli.v1/app.go:482 +0x1e0
go.mozilla.org/sops/vendor/gopkg.in/urfave/cli%2ev1.(*App).Run(0xc420134000, 0xc42007c090, 0x3, 0x3, 0x0, 0x0)
	/Users/samihaahtinen/sources/go/src/go.mozilla.org/sops/vendor/gopkg.in/urfave/cli.v1/app.go:240 +0x596
main.main()
	/Users/samihaahtinen/sources/go/src/go.mozilla.org/sops/cmd/sops/main.go:284 +0xe0e
```

--- a/kms/keysource.go
+++ b/kms/keysource.go
@@ -191,7 +191,7 @@ func ParseKMSContext(in interface{}) map[string]*string {
 		for k, v := range in {
 			value, ok := v.(string)
 			if !ok {
-				fmt.Printf("[WARNING]: KMS Encryption Context contains a non-string value, context will not be used")
+				fmt.Println("[WARNING]: KMS Encryption Context contains a non-string value, context will not be used")
 				return nil
 			}
 			out[k] = &value
@@ -203,12 +203,12 @@ func ParseKMSContext(in interface{}) map[string]*string {
 		for k, v := range in {
 			key, ok := k.(string)
 			if !ok {
-				fmt.Printf("[WARNING]: KMS Encryption Context contains a non-string key, context will not be used")
+				fmt.Println("[WARNING]: KMS Encryption Context contains a non-string key, context will not be used")
 				return nil
 			}
 			value, ok := v.(string)
 			if !ok {
-				fmt.Printf("[WARNING]: KMS Encryption Context contains a non-string value, context will not be used")
+				fmt.Println("[WARNING]: KMS Encryption Context contains a non-string value, context will not be used")
 				return nil
 			}
 			out[key] = &value

--- a/kms/keysource.go
+++ b/kms/keysource.go
@@ -5,7 +5,6 @@ import (
 	"fmt"
 	"os"
 	"regexp"
-	"sort"
 	"strings"
 	"time"
 
@@ -162,8 +161,8 @@ func (key MasterKey) createSession() (*session.Session, error) {
 }
 
 // ToMap converts the MasterKey to a map for serialization purposes
-func (key MasterKey) ToMap() map[string]string {
-	out := make(map[string]string)
+func (key MasterKey) ToMap() map[string]interface{} {
+	out := make(map[string]interface{})
 	out["arn"] = key.Arn
 	if key.Role != "" {
 		out["role"] = key.Role
@@ -171,28 +170,47 @@ func (key MasterKey) ToMap() map[string]string {
 	out["created_at"] = key.CreationDate.UTC().Format(time.RFC3339)
 	out["enc"] = key.EncryptedKey
 	if key.EncryptionContext != nil {
-		var outContexts []string
+		outcontext := make(map[string]string)
 		for k, v := range key.EncryptionContext {
-			outContexts = append(outContexts, k+":"+*v)
+			outcontext[k] = *v
 		}
-		sort.Strings(outContexts)
-		out["context"] = strings.Join(outContexts, ",")
+		out["context"] = outcontext
 	}
 	return out
 }
 
-// ParseKMSContext takes a comma-separated list of KMS context key:value pairs and returns a map
-func ParseKMSContext(in string) map[string]*string {
-	if in == "" {
-		return nil
-	}
+// ParseKMSContext takes either a KMS context map or a comma-separated list of KMS context key:value pairs and returns a map
+func ParseKMSContext(in interface{}) map[string]*string {
 	out := make(map[string]*string)
-	for _, kv := range strings.Split(in, ",") {
-		kv := strings.Split(kv, ":")
-		if len(kv) != 2 {
+
+	switch in := in.(type) {
+	case map[string]interface{}:
+		if len(in) == 0 {
 			return nil
 		}
-		out[kv[0]] = &kv[1]
+		for k, v := range in {
+			var v = v.(string)
+			out[k] = &v
+		}
+	case map[interface{}]interface{}:
+		if len(in) == 0 {
+			return nil
+		}
+		for k, v := range in {
+			var v = v.(string)
+			out[k.(string)] = &v
+		}
+	case string:
+		if in == "" {
+			return nil
+		}
+		for _, kv := range strings.Split(in, ",") {
+			kv := strings.Split(kv, ":")
+			if len(kv) != 2 {
+				return nil
+			}
+			out[kv[0]] = &kv[1]
+		}
 	}
 	return out
 }

--- a/kms/keysource.go
+++ b/kms/keysource.go
@@ -189,16 +189,29 @@ func ParseKMSContext(in interface{}) map[string]*string {
 			return nil
 		}
 		for k, v := range in {
-			var v = v.(string)
-			out[k] = &v
+			value, ok := v.(string)
+			if !ok {
+				fmt.Printf("[WARNING]: KMS Encryption Context contains a non-string value, context will not be used")
+				return nil
+			}
+			out[k] = &value
 		}
 	case map[interface{}]interface{}:
 		if len(in) == 0 {
 			return nil
 		}
 		for k, v := range in {
-			var v = v.(string)
-			out[k.(string)] = &v
+			key, ok := k.(string)
+			if !ok {
+				fmt.Printf("[WARNING]: KMS Encryption Context contains a non-string key, context will not be used")
+				return nil
+			}
+			value, ok := v.(string)
+			if !ok {
+				fmt.Printf("[WARNING]: KMS Encryption Context contains a non-string value, context will not be used")
+				return nil
+			}
+			out[key] = &value
 		}
 	case string:
 		if in == "" {
@@ -207,6 +220,7 @@ func ParseKMSContext(in interface{}) map[string]*string {
 		for _, kv := range strings.Split(in, ",") {
 			kv := strings.Split(kv, ":")
 			if len(kv) != 2 {
+				fmt.Printf("[WARNING]: KMS Encryption Context could not be parsed, context will not be used")
 				return nil
 			}
 			out[kv[0]] = &kv[1]

--- a/kms/keysource_test.go
+++ b/kms/keysource_test.go
@@ -73,6 +73,27 @@ func TestKMSKeySourceFromString(t *testing.T) {
 func TestParseEncryptionContext(t *testing.T) {
 	value1 := "value1"
 	value2 := "value2"
+	// map from YAML
+	var yamlmap = map[interface{}]interface{}{
+		"key1": value1,
+		"key2": value2,
+	}
+	assert.Equal(t, ParseKMSContext(yamlmap), map[string]*string{
+		"key1": &value1,
+		"key2": &value2,
+	})
+	assert.Nil(t, ParseKMSContext(map[interface{}]interface{}{}))
+	// map from JSON
+	var jsonmap = map[string]interface{}{
+		"key1": value1,
+		"key2": value2,
+	}
+	assert.Equal(t, ParseKMSContext(jsonmap), map[string]*string{
+		"key1": &value1,
+		"key2": &value2,
+	})
+	assert.Nil(t, ParseKMSContext(map[string]interface{}{}))
+	// sops 2.0.x formatted encryption context as a comma-separated list of key:value pairs
 	assert.Equal(t, ParseKMSContext("key1:value1,key2:value2"), map[string]*string{
 		"key1": &value1,
 		"key2": &value2,
@@ -87,7 +108,6 @@ func TestParseEncryptionContext(t *testing.T) {
 func TestKeyToMap(t *testing.T) {
 	value1 := "value1"
 	value2 := "value2"
-	value3 := "value3"
 	key := MasterKey{
 		CreationDate: time.Date(2016, time.October, 31, 10, 0, 0, 0, time.UTC),
 		Arn:          "foo",
@@ -96,14 +116,16 @@ func TestKeyToMap(t *testing.T) {
 		EncryptionContext: map[string]*string{
 			"key1": &value1,
 			"key2": &value2,
-			"AAA_this_key_should_be_first": &value3,
 		},
 	}
-	assert.Equal(t, map[string]string{
+	assert.Equal(t, map[string]interface{}{
 		"arn":        "foo",
 		"role":       "bar",
 		"enc":        "this is encrypted",
 		"created_at": "2016-10-31T10:00:00Z",
-		"context":    "AAA_this_key_should_be_first:value3,key1:value1,key2:value2",
+		"context":    map[string]string{
+			"key1": value1,
+			"key2": value2,
+		},
 	}, key.ToMap())
 }

--- a/kms/keysource_test.go
+++ b/kms/keysource_test.go
@@ -83,6 +83,12 @@ func TestParseEncryptionContext(t *testing.T) {
 		"key2": &value2,
 	})
 	assert.Nil(t, ParseKMSContext(map[interface{}]interface{}{}))
+	assert.Nil(t, ParseKMSContext(map[interface{}]interface{}{
+		"key1": 1,
+	}))
+	assert.Nil(t, ParseKMSContext(map[interface{}]interface{}{
+		1: "value",
+	}))
 	// map from JSON
 	var jsonmap = map[string]interface{}{
 		"key1": value1,
@@ -93,6 +99,9 @@ func TestParseEncryptionContext(t *testing.T) {
 		"key2": &value2,
 	})
 	assert.Nil(t, ParseKMSContext(map[string]interface{}{}))
+	assert.Nil(t, ParseKMSContext(map[string]interface{}{
+		"key1": 1,
+	}))
 	// sops 2.0.x formatted encryption context as a comma-separated list of key:value pairs
 	assert.Equal(t, ParseKMSContext("key1:value1,key2:value2"), map[string]*string{
 		"key1": &value1,

--- a/pgp/keysource.go
+++ b/pgp/keysource.go
@@ -201,8 +201,8 @@ func (key *MasterKey) passphrasePrompt(keys []openpgp.Key, symmetric bool) ([]by
 }
 
 // ToMap converts the MasterKey into a map for serialization purposes
-func (key MasterKey) ToMap() map[string]string {
-	out := make(map[string]string)
+func (key MasterKey) ToMap() map[string]interface{} {
+	out := make(map[string]interface{})
 	out["fp"] = key.Fingerprint
 	out["created_at"] = key.CreationDate.UTC().Format(time.RFC3339)
 	out["enc"] = key.EncryptedKey

--- a/sops.go
+++ b/sops.go
@@ -303,7 +303,7 @@ type MasterKey interface {
 	Decrypt() ([]byte, error)
 	NeedsRotation() bool
 	ToString() string
-	ToMap() map[string]string
+	ToMap() map[string]interface{}
 }
 
 // Store provides a way to load and save the sops tree along with metadata
@@ -426,7 +426,7 @@ func (m *Metadata) ToMap() map[string]interface{} {
 	out["mac"] = m.MessageAuthenticationCode
 	out["version"] = m.Version
 	for _, ks := range m.KeySources {
-		var keys []map[string]string
+		var keys []map[string]interface{}
 		for _, k := range ks.Keys {
 			keys = append(keys, k.ToMap())
 		}
@@ -550,7 +550,7 @@ func mapKMSEntriesToKeySource(in []interface{}) (KeySource, error) {
 			return keysource, fmt.Errorf("Could not parse creation date: %s", err)
 		}
 		if _, ok := entry["context"]; ok {
-			key.EncryptionContext = kms.ParseKMSContext(entry["context"].(string))
+			key.EncryptionContext = kms.ParseKMSContext(entry["context"])
 		}
 		key.CreationDate = creationDate
 		keysource.Keys = append(keysource.Keys, key)

